### ISSUE TITLE
TestDiscoveryWatchlabels/Discovery/TestDiscoverySpecPreserved

### ISSUE THREAD

**Issue by bewebi** [solo-io/gloo#10250]
### Which tests failed?

TestDiscoveryWatchlabels/Discovery/TestDiscoverySpecPreserved

Logs:
```
=== RUN   TestDiscoveryWatchlabels/Discovery/TestDiscoverySpecPreserved
service/example-svc created
=== NAME  TestDiscoveryWatchlabels
    status.go:71: 
        Timed out after 20.001s.
        The function passed to Eventually failed at /home/runner/work/gloo/gloo/test/kubernetes/testutils/assertions/status.go:70 with:
        have matcher for namespace discovery-watchlabels-test which is not found
service "example-svc" deleted
```

### Initial Investigation

This has been reproducible locally, though infrequently, whereas it seems to occur much more frequently in CI

This flake occurs when an Upstream created by Discovery does not have its status populated

The failure occurs [here](https://github.com/solo-io/gloo/blob/bc01f4dc028bf0d6724b6607a2420259e748ef6d/test/kubernetes/e2e/features/discovery_watchlabels/discovery_watchlabels_suite.go#L133), on the assertion that eventually an Upstream, which we expect Discovery to create, has an "Accepted" namespaced status

When debugging I found that occasionally the Upstream is created and never gets a status, which leads that assertion to fail

I'm not sure what would cause an Upstream not to get a status, but it may be notable that this failure seems to exclusively occur on the second test, `TestDiscoverySpecPreserved`, and not `TestDiscoverUpstreamMatchingWatchLabels`, suggesting that there may be some form of test pollution

### Additional Information

[One example](https://github.com/solo-io/gloo/actions/runs/11547696038/job/32137980538) of a failure on nightlies

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1207635280052738/1208757360394486) by [Unito](https://www.unito.io)

**Comment by sam-heilbron** [comment]
"I'm not sure what would cause an Upstream not to get a status" - The gloo pod is responsible for writing statuses, and to me a status not getting written is a sign of a potential bug. Even if there is test pollution, I would expect a resource to be processed. 

I would recommend looking to see if we apply another resource (ie trigger translation) does a status appear on this object? If so, it's a sign that the ControlPlane may have "dropped" an event. If not, it's a sign that there's potentially a more fundamental bug (I would look at debug logs of the controlplane to see why that upstream isn't receiving updates)

**Comment by bewebi** [comment]
Indeed, when the test enters this stuck state, applying another resource and re-triggering translation causes a status to be written to the discovered Upstream, unblocking the test

**Comment by sam-heilbron** [comment]
Great! We've cut the problem space down to the [snapshot creation](https://github.com/solo-io/gloo/blob/main/projects/gloo/pkg/api/v1/gloosnapshot/api_snapshot_emitter.sk.go), and confirmed that it's not in the snapshot processing.

The next step I would take is to try to find the area of the code (https://github.com/solo-io/gloo/blob/main/projects/gloo/pkg/api/v1/gloosnapshot/api_snapshot_emitter.sk.go) that is reponsible for snapshot creation and try to find a way to reproduce this in a unit test. The goal (I think) should be to understand how widespread this is. Is it just upstreams that we're missing events on? Is it all resources? Are they skipped (ignored) or dropped (too much throughput)?

Answers to this will inform how impactful we believe this bug is to the product, and thus how much we want to continue investigating in this space

**Comment by bewebi** [comment]
Key finding here:
When the flake occurs we see the following line, from [here](https://github.com/solo-io/gloo/blob/adc0e0e2fb31082ccd86de26517e4e5b5df33de7/projects/gateway2/setup/ggv2setup.go#L202):
```
{"level":"debug","ts":"2024-11-01T20:37:58.067Z","logger":"gloo.v1.event_loop.setup.reporter","caller":"reporter/reporter.go:376","msg":"skipping report for name:\"discovery-watchlabels-test-example-svc-8000\" namespace:\"discovery-watchlabels-test\" as it has not changed","version":"0.0.2-test"}
```
When the test succeeds we see the following line, from [here](https://github.com/solo-io/gloo/blob/adc0e0e2fb31082ccd86de26517e4e5b5df33de7/projects/gateway2/setup/ggv2setup.go#L222):
```
{"level":"debug","ts":"2024-11-01T20:41:31.297Z","logger":"gloo.v1.event_loop.setup.reporter","caller":"reporter/reporter.go:398","msg":"wrote report for name:\"discovery-watchlabels-test-example-svc-8000\" namespace:\"discovery-watchlabels-test\" : state:Accepted reported_by:\"gloo\"","version":"0.0.2-test"}
```

Prior to report writing, we get a log from [this line](https://github.com/solo-io/gloo/blob/ea468b67ab366b4ded96fbc4bda7fe356065cac6/projects/gloo/pkg/syncer/envoy_translator_syncer.go#L210) that prints all resources that will get reports
In the flake scenario the `discovery-watchlabels-test-example-svc-8000` Upstream does already have `namespaced_statuses` populated, which is why `WriteReports()` then skips writing it
In the non-flake scenario, the Upstream doesn’t have `namespaced_statuses` populated, so `WriteReports()` doesn’t skip writing it
This looks like a race of some sort whereby the deletion of the identically-named Upstream from the previous test, hasn’t fully processed by the time `statusClient.GetStatus()` is invoked [here](https://github.com/solo-io/gloo/blob/adc0e0e2fb31082ccd86de26517e4e5b5df33de7/projects/gateway2/setup/ggv2setup.go#L199), so the client returns the status of the previous iteration of the Upstream

This gives us, at the very least, a slightly better sidestep option, which is to ensure the naming is different across specs without weakening the assertion about status
We still need to investigate ways to eliminate the race condition

To that end, I think we’ll need to determine a detailed order of events
Based on my debug logging we have the same number of events written to/by the Upstream watcher when running the suite, however the last event when we have a flake includes the status and when there is no flake it does not
What I don’t understand is:
- What triggers the event if the state of the world still has the Upstream with the status populated (ie before delete)?
- Regardless, why doesn’t the (deletion and subsequent) creation of an Upstream trigger a new event?

My understanding is that the watch is really a periodic GET request to the k8s API server and a comparison with the last known state of the world, generating an event if there is a diff (I was struggling to trace the exact period, but I think it may be 1s)
When the internal state of the world (ie cache) says there is a status but etcd has no status this seemingly should trigger an event every second until there is no diff

**Comment by sheidkamp** [comment]
1.18 PR: https://github.com/solo-io/gloo/actions/runs/11790577469/job/32841256480?pr=10284

**Comment by ashishb-solo** [comment]
failed again in https://github.com/solo-io/solo-projects/actions/runs/11831088797/job/32965702334

**Comment by npolshakova** [comment]
Happened on 1.17.x nightlys: https://github.com/solo-io/solo-projects/actions/runs/12113623368

**Comment by ryanrolds** [comment]
Sounds like the fix needs backported.

**Comment by bewebi** [comment]
Occurred on main OSS nightly: https://github.com/k8sgateway/k8sgateway/actions/runs/12193272104/job/34015253164

**Comment by ryanrolds** [comment]
Do we think it's a new issue or the earlier fix didn't resolve it? My only understanding is that a fix made prior to this ticket being assigned to me.

**Comment by bewebi** [comment]
What fix was implemented? This failure was on `main` so backporting shouldn't be relevant

**Comment by ryanrolds** [comment]
My understanding is we fix was implemented in 1.18. I ran the test 100 times locally and didn't see it fail. I will investigate further later today. Is this a blocker?

**Comment by bewebi** [comment]
I don't believe so, just reporting flakes from nightly runs

**Comment by ryanrolds** [comment]
It will get my attention early next week.

**Comment by ryanrolds** [comment]
Dropping this. I'm not on-call and didnt get to it.

**Comment by ryanrolds** [comment]
Looks like someone needs to remove me from this.

**Comment by jenshu** [comment]
closing, not relevant in kgateway.
moved to https://github.com/solo-io/gloo/issues/10678
